gh-100226: Clarify StreamReader.read behavior (#101807)
python/cpython: Doc/library/asyncio-stream.rst

   .. coroutinemethod:: read(n=-1)

>       Read up to *n* bytes.

Reviewer: ```suggestion
      Read up to *n* bytes from the stream.
```

For explicitness, and to match the docstring version

      If *n* is not provided, or set to ``-1``,

>       read until EOF. After EOF is received, return all read bytes.

Reviewer: ```suggestion
      If *n* is not provided or set to ``-1``,
      read until EOF (``EOT``, ``b"\x04"``/:kbd:`Ctrl-D`) is received,
      then return all read :class:`bytes`.
```

* Clarify that in this context (at least, I'm assuming this is the case if this follows the Unix `read` semantics like Guido indicated), EOF is specifically the `EOT` character code, rather than `SUB` (`\x1A`/`Ctrl-Z`, the Windows EOF), some other byte or some other out of band signalling.
* Join the sentences accordingly to avoid repetition the same
* Be more explicit and cross-reference the returned data type
* Elide unnecessary internal comma

Author: I'm actually unsure whether the details about the character code are correct. I suppose the standard use case of the `StreamReader` is to read from a TCP socket. It seems to me that this maps to the system's `recv` function (though, I didn't look through all details). The man page on a Debian Linux says:

> When a stream socket peer has performed an orderly shutdown, the return value will be 0 (the traditional "end-of-file" return).

Similarly, the man page on my macOS says:

> If no messages are available to be received and the peer has performed an orderly shutdown, the value 0 is returned.

This sounds to me like there is no actual character code read. (I suppose a TCP FIN packet will be received, which is handled by the network stack and causes read just to return 0.)

I suppose for a different kind of stream (file instead of socket?), a character code might be used to indicate EOF, but at that point it would become a detail of the underlying stream and be no longer a general aspect of the `StreamReader`.

But I might also be misunderstanding things? I'm not that familiar with these low-level details.

(The other changes look good to me btw.)

Reviewer 2: “EOF” refers to the concept of ending the stream, not to a character. ^D and others are not passed to the application (except maybe in raw tty mode) and should not be mentioned here.

      If EOF was received and the internal buffer is empty,
      return an empty ``bytes`` object.

>       If *n* is zero, return empty ``bytes`` object immediately.

Reviewer: ```suggestion
      If *n* is ``0``, return an empty ``bytes`` object immediately.
```

* Fix grammar
* Use the appropriate literal

>       If *n* is positive, this function tries to read *n* bytes, and may return
>       less or equal bytes than requested, but at least one byte. If EOF was
>       received before any byte is read, this function returns an empty
>       ``bytes`` object.

Reviewer: ```suggestion
      If *n* is positive, read at most *n* bytes.
      May return less than *n* bytes, if EOF is reached first.
      If EOF is received before any byte is read,
      return an empty ``bytes`` object.
```

* Specify the actual reason for less than *n* bytes being returned—EOF is reached first (assuming I'm interpreting this correctly, and it parallels the ``-1`` case).
* Avoid apparent contradictory statement that at least one byte will be returned, yet can also return empty bytes if EOF is received first
* Split first sentence into two, for clarity and readability
* Elide redundant (and imprecise) "this function" verbiage, and be consistent with the phrasing used elsewhere here

Author: > Specify the actual reason for less than n bytes being returned—EOF is reached first (assuming I'm interpreting this correctly, and it parallels the -1 case).

This is not correct. It can also return less than `n` bytes, if all `n` bytes are not immediately available (but at least 1 byte). This was the cause of a bug in one of my Python programs (less than `n` bytes because EOF would have been handled fine).

Reviewer 2: Indeed.

Author: I pushed which I hope incorporates your suggestions but keeps the correct reasons less than `n` bytes might be returned.

Reviewer: I've addressed further refinements in a followup suggestion, which was applied.

   .. coroutinemethod:: read(n=-1)

      Read up to *n* bytes from the stream.

>       If *n* is not provided, or set to ``-1``,

Reviewer: ```suggestion
      If *n* is not provided or set to ``-1``,
```

This comma isn't necessary, and number of commas in this sentence harms readability and clear hierarchy

>       read until EOF, then return all read :class:`bytes`.

Reviewer: As a beginning reader, before having read @jgosmann 's [detailed explaination](https://github.com/python/cpython/pull/101807#discussion_r1108985811) and @gvanrossum 's summary, I was evidently confused here about what "until EOF" specifically meant in this context. It therefore might warrant some more clarity here, but maybe its just my lack of subject matter expertise compared with the target audience.

      If EOF was received and the internal buffer is empty,
      return an empty ``bytes`` object.

      If *n* is ``0``, return an empty ``bytes`` object immediately.

      If *n* is positive, wait for at least 1 byte to become available
      (or EOF), then return the available ``bytes`` which might be less

>       than *n*.

Reviewer 2: if there's still data in the buffer, this doesn't wait. Maybe the text should allude to that?

Reviewer: Yeah, I've gone ahead and included that in my suggestion

>       If *n* is positive, wait for at least 1 byte to become available
>       (or EOF), then return the available ``bytes`` which might be less
>       than *n*.

Reviewer: Unlike the previous, this appears to neglect to explicitly mention anything about what the value of `n` actually does here (the number of bytes to return, at most). Adding that back, eliding the now unnecessary mention that fewer bytes may be returned (since that's what "at most" means), implementing Guido's suggestion and fixing a couple other things, we have the much simplified

```suggestion
      If *n* is positive, return at most *n* available ``bytes``
      as soon as at least 1 byte is available in the internal buffer.
```
